### ISSUE TITLE
Delay Node when set to Rate Limit mode delays messages for topics

### ISSUE THREAD

**Issue by Slyke** [node-red/node-red#3260]
### Current Behavior

If setting the action from "All messages" to "For each msg.topic" the message is not forwarded immediately when it should be. Messages that are queued eventually get set free, but no subsequent messages are ever sent to the second output if "Send intermediate messages on 2nd output". It only ever allows one message to be queued regardless of what the X messages per T period is.

### Expected Behavior

Message should be forwarded immediately, but prevent subsequent messages being forwarded until the limit is cleared.

### Steps To Reproduce

Click the inject nodes in the example flow. I've setup 2 flows, the top one with topic rate filtering and the bottom one with filtering on all messages. You'll see the top one is always delayed.

### Example flow

```
[{"id":"b1dc1d46e5c77129","type":"tab","label":"Flow 1","disabled":false,"info":"","env":[]},{"id":"2112ad7af5dddc95","type":"inject","z":"b1dc1d46e5c77129","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payloadType":"date","x":140,"y":120,"wires":[["ef18a9904dd4d751"]]},{"id":"bc5c04e22e5d8680","type":"debug","z":"b1dc1d46e5c77129","name":"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"false","statusVal":"","statusType":"auto","x":870,"y":140,"wires":[]},{"id":"f5ccdf992407db39","type":"delay","z":"b1dc1d46e5c77129","name":"","pauseType":"timed","timeout":"5","timeoutUnits":"seconds","rate":"1","nbRateUnits":"30","rateUnits":"second","randomFirst":"1","randomLast":"5","randomUnits":"seconds","drop":true,"allowrate":false,"outputs":2,"x":560,"y":160,"wires":[["bc5c04e22e5d8680"],["776dab2f877c6d3e"]]},{"id":"ef18a9904dd4d751","type":"function","z":"b1dc1d46e5c77129","name":"","func":"msg.topic = 'top inject';\nmsg.payload = msg.topic;\nreturn msg;","outputs":1,"noerr":0,"initialize":"","finalize":"","libs":[],"x":340,"y":120,"wires":[["f5ccdf992407db39"]]},{"id":"cfff25186bd645d6","type":"debug","z":"b1dc1d46e5c77129","name":"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"false","statusVal":"","statusType":"auto","x":970,"y":200,"wires":[]},{"id":"25e3d637649c1eb3","type":"inject","z":"b1dc1d46e5c77129","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"","payloadType":"date","x":140,"y":180,"wires":[["35b3b90b912ed239"]]},{"id":"35b3b90b912ed239","type":"function","z":"b1dc1d46e5c77129","name":"","func":"msg.topic = 'bottom inject';\nmsg.payload = msg.topic;\nreturn msg;","outputs":1,"noerr":0,"initialize":"","finalize":"","libs":[],"x":340,"y":180,"wires":[["f5ccdf992407db39"]]},{"id":"776dab2f877c6d3e","type":"function","z":"b1dc1d46e5c77129","name":"","func":"msg.payload = 'dropped ' + msg.payload;\nreturn msg;","outputs":1,"noerr":0,"initialize":"","finalize":"","libs":[],"x":780,"y":200,"wires":[["cfff25186bd645d6"]]},{"id":"858d86099b07157b","type":"inject","z":"b1dc1d46e5c77129","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payloadType":"date","x":140,"y":380,"wires":[["5a914ff9b6439ea6"]]},{"id":"64faae948c4f6954","type":"debug","z":"b1dc1d46e5c77129","name":"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"false","statusVal":"","statusType":"auto","x":870,"y":400,"wires":[]},{"id":"7a837a46cc6682b9","type":"delay","z":"b1dc1d46e5c77129","name":"","pauseType":"rate","timeout":"5","timeoutUnits":"seconds","rate":"1","nbRateUnits":"30","rateUnits":"second","randomFirst":"1","randomLast":"5","randomUnits":"seconds","drop":true,"allowrate":false,"outputs":2,"x":540,"y":420,"wires":[["64faae948c4f6954"],["6e945d20cdb46372"]]},{"id":"5a914ff9b6439ea6","type":"function","z":"b1dc1d46e5c77129","name":"","func":"msg.topic = 'top inject';\nmsg.payload = msg.topic;\nreturn msg;","outputs":1,"noerr":0,"initialize":"","finalize":"","libs":[],"x":340,"y":380,"wires":[["7a837a46cc6682b9"]]},{"id":"ccc8e74e0a7e4277","type":"debug","z":"b1dc1d46e5c77129","name":"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"false","statusVal":"","statusType":"auto","x":970,"y":460,"wires":[]},{"id":"d1fd331d3608d441","type":"inject","z":"b1dc1d46e5c77129","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"","payloadType":"date","x":140,"y":440,"wires":[["faf0c3f4167b3d96"]]},{"id":"faf0c3f4167b3d96","type":"function","z":"b1dc1d46e5c77129","name":"","func":"msg.topic = 'bottom inject';\nmsg.payload = msg.topic;\nreturn msg;","outputs":1,"noerr":0,"initialize":"","finalize":"","libs":[],"x":340,"y":440,"wires":[["7a837a46cc6682b9"]]},{"id":"6e945d20cdb46372","type":"function","z":"b1dc1d46e5c77129","name":"","func":"msg.payload = 'dropped ' + msg.payload;\nreturn msg;","outputs":1,"noerr":0,"initialize":"","finalize":"","libs":[],"x":780,"y":460,"wires":[["ccc8e74e0a7e4277"]]},{"id":"882b5f6ffef6c419","type":"comment","z":"b1dc1d46e5c77129","name":"Sends message immediately","info":"","x":540,"y":340,"wires":[]},{"id":"d3bdf4f53c494af0","type":"comment","z":"b1dc1d46e5c77129","name":"Messages are delayed","info":"","x":520,"y":80,"wires":[]}]
```

### Environment

- Node-RED version: 2.1.3
- Node.js version:
- npm version:
- Platform/OS: Kubernetes/Linux
- Browser: Chrome

**Comment by dceejay** [comment]
This is half working as designed. 
When in by topic mode there are also two sub options - to release each topic in turn - or to release all topics at once. So to accommodate those there is simply a timer loop that ticks every x seconds... and it has to wait for the time slot in order for any other topics to arrive, and it doesn't matter when the messages arrive within that time - if a new message arrives on the topic then it replaces the one a the head of the queue - so the one released is always the most recent for that topic. This is the desired behaviour as it allows the most recent message to be send and old data to be discarded.

What isn't happening  - is that  when in this mode - currently nothing ever gets sent to the second output... so what would make sense in this case ? if a new message arrives on the topic then in order to maintain the existing behaviour, then the "old" message should probably get pushed to the second output as it is being "discarded".

**Comment by Slyke** [comment]
How long is that timer? I set the time thresholds to 30s and it was taking about that long for the message to be released. I could see the node had a number under it for how many message topics it had in its buffer.

When not in "for each topic" mode the messages are forwarded immediately. I would expect the first message(s) to go through, all following messages that exceed the threshold should be sent to the second output, or dropped. This way it works the same as when not in "for each topic" mode.

It just appears like when in the "for each topic" mode it's activating the delay logic and rate limit logic, when it should only be activating the rate limit logic.

**Comment by Slyke** [comment]
I've been going through the code, but haven't found where exactly the issue is coming up. When I get some more time to investigate, I'll keep looking: https://github.com/node-red/node-red/blob/master/packages/node_modules/%40node-red/nodes/core/function/89-delay.js#L227

**Comment by dceejay** [comment]
sort of but not quite.... the first message to arrive is what starts the timer ticking... 
https://github.com/node-red/node-red/blob/master/packages/node_modules/%40node-red/nodes/core/function/89-delay.js#L346
so if say set to 10 secs then yes first one out is sent at 10 secs.... if another arrives at the 15 second mark it will get released at 20 secs - or if at 19 - still at 20 secs. And then the next lot at 30 secs. I can see there is an argument for releasing the first right away when in "send each topic in turn" mode as that first message will be the first one sent - but in "all at once" mode it really does need to wait to accumulate potentially several topics. 

However the overall effect is the same the rate is limited to a number of messages per time interval. The whole purpose of these modes was to try to evenly space out a bunch of messages when using a low bandwidth comms link - so as to only send/sample a small amount of data from a stream and only to send the latest topic of each down a slow link. Once started the spacing would try to continue at an even pace so as not to overwhelm the data link.

**Comment by Slyke** [comment]
How come it sends it out immediately when "All messages" is selected, but is delayed for "For each msg.topic", even though they are both under "Rate Limit"? It doesn't really imply that there should be any delay when "Send all topics" is set, and especially when "Send intermediate messages on 2nd output" is set. Delaying a message can still be achieved with the same node, but set in delay mode and connecting them in series, but topics can be rate limited without delay, and instead just be discarded, or sent on the second output.

My use case is that I send an alert to Discord when my garage, front or back door is open via MQTT on an Envisalink module from my home security system. The way that it reports it is that the security system polls the doors last detected closed time in `seconds ago`, which is continuously reset every 5 to 10 seconds if the door is open as the security system polls it. This causes a lot of spam notifications. So I put in rate limiting for the messages which works fine and lets me know immediately if the door is opened, but will only send a few messages a minute for all doors. I instead want to have it limit the messages sent per door, which is why I want to use each door as a topic.

So the problem is, when using each door as a topic, it only forwards the message once the time period is up. When setting an upper limit of messages per day, that means that it'd wait a day before it sent an alert.

I hope that what I wrote makes sense haha.

![image](https://user-images.githubusercontent.com/6442613/141761129-73653e94-2616-41b5-8d5f-fc6398d66637.png)

**Comment by dceejay** [comment]
to de-duplicate repeated messages you would be much better off using the RBE/Filter node which does excaly that and also does it per topic.

**Comment by Slyke** [comment]
> to de-duplicate repeated messages you would be much better off using the RBE/Filter node which does exactly that (only sends changes of state) - and also does it per topic.

Yep, this is what I want, but I only want it to block if that particular topic has been sent too many times in T amount of time and then send the rejected messages to a different output. The RBE node blocks unless it has changed in value. I actually originally used this node. I then switched to the delay node, which does exactly what I want with the rate limiting, except when setting it to check on a per topic basis.

I spose adding another property that has the time diff since the last alert for that topic and current time on the RBE node could also work, but it's getting more complex.

The other solution I thought of was just to have the `DiscordMap Zone to Location` function node just have several outputs (one for each topic), but that is not scalable and becomes unmanagable.

I think the best solution would be to have the rate limit not buffer any messages when set to per topic, like it does when set to per message. And instead forward it to either output, or just discard if only one output.

**Comment by dceejay** [comment]
the RBE node can be reset every x seconds - so you can send a burst of different topics through - block any further repeats... then reset it and resend the next batch ?

**Comment by Slyke** [comment]
An RBE node that blocks on a per topic basis, that's fed a reset by a delay node might work, but then all those other messages are dropped - they're not sent to the second output like they are on the delay node when set to rate limit.

![image](https://user-images.githubusercontent.com/6442613/141779495-eaf5ffe9-391d-4ec6-a930-662ecf87cb5e.png)

Those messages without the `Alert @Slyke` would never be sent, if that makes sense. I'm still having trouble understanding why the rate limit only has a delay when set to `For each msg.topic` when it doesn't when set to all messages.

**Comment by dceejay** [comment]
because when set to send all topics that arrive it has to wait for them to arrive in order to  gather them together. When ignoring the topic (all messages) then it can make the decision right away whether to send or drop. (The second output was only recently added so the original choice was only forward or drop.)

What use are those repeated messages ?

**Comment by Slyke** [comment]
Every 15-30 or so seconds all security devices are repolled and the open one's timer are reset to 0 by the security controller box. If the reported timer for that particular security device (door, sensor etc) is less than the amount saved in the flow's context, then NodeRed knows that security device was opened. Since it repolls every 30 seconds, it would spam if a security device is left open (like a front door).

To avoid the spam, I made it only does the `@ alerts` every so often, as set with the rate limiting node, with the rejected messages just being logged without the `@ alert` so that it can still be checked up upon. But it does this for all messages. I was hoping to do this on a per security device basis, using the device ID as the topic for the message.

This way if say the front door was left open for an hour and the backdoor was opened, it would send an `@ alert` for the backdoor.

For context, 3 houses on my street were broken into in the last 2 weeks, so I'm setting up cameras and connecting into my security system so can monitor and be alerted if I or my gf are sleeping or out etc.

**Comment by dceejay** [comment]
so then I'm not understanding why a simple RBE won't do what you want from an external point of view - by all means log all events internally for audit or review - but the RBE would only report each change (door closed -> open) once per change without spamming, and would only send the topic that changed to external rather than all. You can always store all the topics in a context variable if you want to later grab a snapshot of the current state at any time.

**Comment by Slyke** [comment]
Thank you!

**Comment by knolleary** [comment]
This was fixed in 2.1.4
